File: apps/orchestrator-api/src/memory_retriever.py
from typing import Callable, Optional

from src.postgres_client import PostgresClient
from src.qdrant_client import QdrantClient
from src.logging import get_logger

logger = get_logger("memory-retriever")

EmbedFn = Callable[[str], list[float]]


class MemoryRetriever:
    def __init__(self, postgres: PostgresClient, qdrant: QdrantClient, embed_fn: EmbedFn):
        self.postgres = postgres
        self.qdrant = qdrant
        self.embed_fn = embed_fn

    async def retrieve(self, query: dict) -> list[dict]:
        text_query = query.get("query", "")
        categories = query.get("categories")
        limit = query.get("limit", 10)
        min_importance = query.get("min_importance", 1)

        if text_query and categories:
            return await self._hybrid_search(text_query, categories, limit, min_importance)
        elif categories:
            return await self._category_search(categories[0], limit, min_importance)
        elif text_query:
            return await self._vector_search(text_query, limit, min_importance)
        else:
            return []

    async def _hybrid_search(self, query: str, categories: list[str], limit: int, min_importance: int) -> list[dict]:
        try:
            embedding = await self.embed_fn(query)
            filter_payload = {"category": categories[0]} if len(categories) == 1 else None
            vector_results = await self.qdrant.search("memories", embedding, limit=limit * 2, filter_payload=filter_payload)

            seen_ids = set()
            results = []
            for r in vector_results:
                if r["id"] not in seen_ids and r.get("importance", 0) >= min_importance:
                    seen_ids.add(r["id"])
                    results.append(r)
                    if len(results) >= limit:
                        break

            if len(results) < limit:
                for cat in categories:
                    pg_results = await self.postgres.search_by_category(cat, limit=limit - len(results), min_importance=min_importance)
                    for r in pg_results:
                        if r["id"] not in seen_ids:
                            seen_ids.add(r["id"])
                            results.append(r)
                            if len(results) >= limit:
                                break
                    if len(results) >= limit:
                        break

            return results[:limit]
        except Exception as e:
            logger.warn("Hybrid search failed, falling back to category search", {"error": str(e)})
            return await self._category_search(categories[0], limit, min_importance)

    async def _category_search(self, category: str, limit: int, min_importance: int) -> list[dict]:
        return await self.postgres.search_by_category(category, limit, min_importance)

    async def _vector_search(self, query: str, limit: int, min_importance: int) -> list[dict]:
        try:
            embedding = await self.embed_fn(query)
            results = await self.qdrant.search("memories", embedding, limit=limit)
            return [r for r in results if r.get("importance", 0) >= min_importance][:limit]
        except Exception as e:
            logger.warn("Vector search failed", {"error": str(e)})
            return []

File: packages/memory/src/memory-retriever.ts
import { MemoryRetrievalRequest, MemoryRecord, VectorSearchResult, MemoryCategory } from "./types";
import { PostgresClient } from "./postgres-client";
import { QdrantClient } from "./qdrant-client";
import { get_logger } from "./logging";

const logger = get_logger("memory-retriever");

interface RetrievalScoredResult {
  record: MemoryRecord;
  vectorScore: number;
  keywordScore: number;
  importanceWeight: number;
  finalScore: number;
}

export class MemoryRetriever {
  private postgres: PostgresClient;
  private qdrant: QdrantClient;
  private embedFn: (text: string) => Promise<number[]>;

  constructor(postgres: PostgresClient, qdrant: QdrantClient, embedFn: (text: string) => Promise<number[]>) {
    this.postgres = postgres;
    this.qdrant = qdrant;
    this.embedFn = embedFn;
  }

  async retrieve(request: MemoryRetrievalRequest): Promise<MemoryRecord[]> {
    const { query, categories, limit, min_importance, include_embeddings } = request;

    const vectorResults = include_embeddings ? await this.vectorSearch(query, categories, limit * 2) : [];
    const keywordResults = await this.keywordSearch(query, categories, limit * 2);

    const merged = this.mergeAndScore(vectorResults, keywordResults, min_importance);
    const sorted = merged.sort((a, b) => b.finalScore - a.finalScore);

    return sorted.slice(0, limit).map((r) => r.record);
  }

  async retrieveByCategory(category: MemoryCategory, limit: number = 10): Promise<MemoryRecord[]> {
    return this.postgres.getByCategory(category, limit);
  }

  async retrieveRecent(limit: number = 20): Promise<MemoryRecord[]> {
    return this.postgres.getRecentAcrossCategories(limit);
  }

  async retrieveImportant(limit: number = 10): Promise<MemoryRecord[]> {
    const allCategories: MemoryCategory[] = ["identity", "relationship", "lore", "project", "episodic", "technical"];
    const results: MemoryRecord[] = [];

    for (const category of allCategories) {
      const records = await this.postgres.getByCategory(category, limit);
      results.push(...records.filter((r) => r.importance >= 4));
    }

    return results.sort((a, b) => b.importance - a.importance).slice(0, limit);
  }

  private async vectorSearch(
    query: string,
    categories?: MemoryCategory[],
    limit: number = 20,
  ): Promise<VectorSearchResult[]> {
    try {
      const embedding = await this.embedFn(query);
      return this.qdrant.search(embedding, limit, categories);
    } catch (error) {
      logger.error("Vector search failed, falling back to keyword", { error: String(error) });
      return [];
    }
  }

  private async keywordSearch(
    query: string,
    categories?: MemoryCategory[],
    limit: number = 20,
  ): Promise<MemoryRecord[]> {
    const targets = categories ?? ["identity", "relationship", "lore", "project", "episodic", "technical"];
    const results: MemoryRecord[] = [];

    for (const category of targets) {
      const matches = await this.postgres.searchByTitle(category, query, Math.ceil(limit / targets.length));
      results.push(...matches);
    }

    return results;
  }

  private mergeAndScore(
    vectorResults: VectorSearchResult[],
    keywordResults: MemoryRecord[],
    minImportance: number,
  ): RetrievalScoredResult[] {
    const recordMap = new Map<number, RetrievalScoredResult>();

    for (const vr of vectorResults) {
      const memoryId = Number(vr.payload.memory_id);
      if (isNaN(memoryId)) continue;

      const record: MemoryRecord = {
        id: memoryId,
        category: (vr.payload.category as MemoryCategory) ?? "episodic",
        title: String(vr.payload.title ?? ""),
        content: String(vr.payload.content ?? ""),
        importance: Number(vr.payload.importance ?? 3),
        metadata: (vr.payload.metadata as Record<string, unknown>) ?? {},
        created_at: String(vr.payload.created_at ?? ""),
        updated_at: String(vr.payload.updated_at ?? ""),
        last_accessed_at: (vr.payload.last_accessed_at as string | null) ?? null,
        access_count: Number(vr.payload.access_count ?? 0),
      };

      if (record.importance < minImportance) continue;

      recordMap.set(memoryId, {
        record,
        vectorScore: vr.score,
        keywordScore: 0,
        importanceWeight: record.importance / 5,
        finalScore: 0,
      });
    }

    for (const kr of keywordResults) {
      if (kr.importance < minImportance) continue;

      const existing = recordMap.get(kr.id);
      if (existing) {
        existing.keywordScore = 0.8;
        existing.finalScore = this.calculateFinalScore(existing);
      } else {
        const entry: RetrievalScoredResult = {
          record: kr,
          vectorScore: 0,
          keywordScore: 0.8,
          importanceWeight: kr.importance / 5,
          finalScore: 0,
        };
        entry.finalScore = this.calculateFinalScore(entry);
        recordMap.set(kr.id, entry);
      }
    }

    for (const [, entry] of recordMap) {
      entry.finalScore = this.calculateFinalScore(entry);
    }

    return Array.from(recordMap.values());
  }

  private calculateFinalScore(result: RetrievalScoredResult): number {
    const vectorWeight = 0.4;
    const keywordWeight = 0.3;
    const importanceWeight = 0.3;

    return (
      result.vectorScore * vectorWeight +
      result.keywordScore * keywordWeight +
      result.importanceWeight * importanceWeight
    );
  }
}
